Create a page for OpenFOAM (#17)
ACRC/HPC-community-docs: source/engineering/OpenFOAM/multi-node.md

```{code-block} bash
---
linenos: true
---
#!/bin/bash
## Submission script for Cluster
#SBATCH --job-name=my_job
#SBATCH --time=0-01:00:00 # hh:mm:ss
#SBATCH --nodes=4
#SBATCH --ntasks-per-node=28
#SBATCH --cpus-per-task=1

## system error message output file
## leave %j as it's being replaced by JOB ID number
#SBATCH -e my_job_%j.err

## system message output file
#SBATCH -o my_job_%j.out

# record some potentially useful details about the job: 
echo Running on host $(hostname) 
echo Time is $(date) 
echo Directory is $(pwd) 
echo Slurm job ID is ${SLURM_JOBID} 
echo This jobs runs on the following machines: 
echo ${SLURM_JOB_NODELIST} 
printf "\n\n\n\n" 

# Load modules required for runtime e.g.
module load apps/openfoam/6

# Run the solver. Take pisoFoam with 112 processors for example:
mpirun -np 112 pisoFoam -parallel 

echo End Time is $(date) 
echo "Done pisoFoam finish"
printf "\n\n"

```

```{note}
For further information on the structure and syntax of Slurm job scripts, see the [ACRC HPC documentation pages on job types](https://www.acrc.bris.ac.uk/protected/hpc-docs/job_types/index.html).
```
### How to use

> Save the script file in the [case folder](https://www.openfoam.com/documentation/user-guide/2-openfoam-cases), and submit it to Slurm after the decomposition of mesh and initial field data. The number of MPI processes to request is the number of nodes times the number of tasks per node. 

Reviewer: ```suggestion
Save the script file in the [case folder](https://www.openfoam.com/documentation/user-guide/2-openfoam-cases), `cd` to the case folder, and submit the script to Slurm after the decomposition of mesh and initial field data. The number of MPI processes to request is the number of nodes times the number of tasks per node. 
```

---
title: "Multi-Node Jobs"
---

Large OpenFOAM jobs can be run on multi nodes.

## Job Script Template

Here is an example job submission script for a multi-node OpenFOAM job:

```{code-block} bash
---
linenos: true
---
#!/bin/bash
## Submission script for Cluster
#SBATCH --job-name=my_job
#SBATCH --time=0-01:00:00 # hh:mm:ss
#SBATCH --nodes=4
#SBATCH --ntasks-per-node=28
#SBATCH --cpus-per-task=1

> #SBATCH --partition=cpu

Reviewer: ```suggestion
```

There should be no need to specify the `cpu` partition here, as this is the default partition on BC4.

## system error message output file
## leave %j as it's being replaced by JOB ID number
#SBATCH -e aai_LD_%j.err

## system message output file
#SBATCH -o aai_LD_%j.out

# record some potentially useful details about the job: 
echo Running on host $(hostname) 
echo Time is $(date) 
echo Directory is $(pwd) 
echo Slurm job ID is ${SLURM_JOBID} 
echo This jobs runs on the following machines: 
echo ${SLURM_JOB_NODELIST} 
printf "\n\n\n\n" 

# Load modules required for runtime e.g.
module load apps/openfoam/6

#export I_MPI_PROCESS_MANAGER=mpd

> 

Reviewer: Should the script `cd` to the case folder before running `pisoFoam`?

Author: Yes, as different cases usually use different solvers, it would be better that each case has its own script.

# Run the solver. Take pisoFoam with 112 processors for example:
mpirun -np 112 pisoFoam -parallel 

echo End Time is $(date) 

> echo "Done pimpleFoam finish"

Reviewer: ```suggestion
echo "Done pisoFoam finish"
```

printf "\n\n"

```

```{note}
Further explanation of the script could be found in: https://www.acrc.bris.ac.uk/protected/hpc-docs/job_types/index.html
```
### How to use

> Save the script file in the case folder, and submit it to Slurm after the decomposition of mesh and initial field data. The processors number is the nodes number times the number of tasks per node. 

Reviewer: It might be useful to link to some documentation on the case folder.

```suggestion
Save the script file in the [case folder](https://www.openfoam.com/documentation/user-guide/2-openfoam-cases), and submit it to Slurm after the decomposition of mesh and initial field data. The number of MPI processes to request is the number of nodes times the number of tasks per node. 
```

> Save the script file in the case folder, and submit it to Slurm after the decomposition of mesh and initial field data. The processors number is the nodes number times the number of tasks per node. 

Reviewer: What does "after the decomposition of mesh and initial field data" mean here? Do you mean that you need to set up the mesh and initial field data files in the case folder?
